Backend/utils/cronJobRecommendation.js

```diff
@@ -24,50 +38,70 @@ async function calculateRecommendations() {
     for (const song of songs) {
         const { userId, genre, mood, tempo } = song;
         if (!userTagHistory[userId]) {
-            userTagHistory[userId] = [];
+            userTagHistory[userId] = []; // Initialize user's tag history if not already
         }
 
         // Assign numerical values to tags
-        const tagCombination = [
-            tagValues.genre[genre],
-            mood ? tagValues.mood[mood] : 0,
-            tempo ? tagValues.tempo[tempo] : 0
-        ];
+        const tagCombination = {
+            genre: tagValues.genre[genre], // Convert genre to number
+            mood: mood ? tagValues.mood[mood] : 0, // Convert mood to number, if exists
+            tempo: tempo ? tagValues.tempo[tempo] : 0 // Convert tempo to number, if exists
```

Comment:
so is genre required while mood and tempo are optional tags because they can be null? wouldn't it be easier for your algorithm to make all of them required?

Reply:
Yes, as of right now the "genre" tag is the only tag that is required but, to maintain consistency, I'm going to update the code to make all the tags required.

Reply:
Just realized this fix is going to be a big one since I have to make changes from the frontend to the database not just the backend but I'll make sure to circle back to this once I've finished updating the algorithm.